google/cloud/spanner_dbapi/transaction_helper.py

```diff
@@ -61,67 +61,106 @@ def reset(self):
+            result_details = _get_statement_result_checksum(result_rows)
             if is_fetch_all:
-                self._last_statement_result_details = FetchStatement(
-                    statement_type=CursorStatementType.FETCH_ALL,
-                    result_details=result_details,
-                )
+                statement_type = CursorStatementType.FETCH_ALL
+                size = None
             else:
-                self._last_statement_result_details = FetchStatement(
-                    statement_type=CursorStatementType.FETCH_MANY,
-                    result_details=result_details,
-                    size=len(rows),
-                )
-            self._statement_result_details_list.append(
-                self._last_statement_result_details
+                statement_type = CursorStatementType.FETCH_MANY
+                size = len(result_rows)
+
+            last_statement_result_details = FetchStatement(
+                cursor=cursor,
+                statement_type=statement_type,
+                result_type=ResultType.CHECKSUM,
+                result_details=result_details,
+                size=size,
             )
+            self._last_statement_details_per_cursor[
+                cursor
+            ] = last_statement_result_details
+            self._statement_result_details_list.append(last_statement_result_details)
 
     def add_execute_statement_for_retry(
-        self, parsed_statement, sql, args, rowcount, exception, is_execute_many
+        self, cursor, sql, args, exception, is_execute_many
     ):
+        """
+        StatementDetails to be added to _statement_result_details_list whenever execute or
+        executemany method is called on the cursor.
+
+        :param cursor: original Cursor object on which statement executed in the transaction
+        :param sql: Input param of the execute/executemany method
+        :param args: Input param of the execute/executemany method
+        :param exception: Not none in case non-aborted exception is thrown on the original
+        statement execution
+        :param is_execute_many: True in case of executemany statement execution
+        """
         if not self._connection._client_transaction_started:
             return
         statement_type = CursorStatementType.EXECUTE
         if is_execute_many:
             statement_type = CursorStatementType.EXECUTE_MANY
+
+        result_type = ResultType.NONE
         result_details = None
         if exception is not None:
+            result_type = ResultType.EXCEPTION
             result_details = exception
-        elif (
-            parsed_statement.statement_type == StatementType.INSERT
-            or parsed_statement.statement_type == StatementType.UPDATE
-            or parsed_statement.client_side_statement_type
-            == ClientSideStatementType.RUN_BATCH
-        ):
-            result_details = rowcount
+        elif cursor._batch_dml_rows_count is not None:
```

Comment:
One very special thing about batch DML is that it can contain __both__ update counts __and__ an exception. If you for example execute a batch DML statement like this:

```sql
start batch dml;
insert into test (id, value) values (1, 'One');
insert into test (id, value) values (2, 'Two');
insert into test (id, value) values (1, 'Three');
```

Then the result will be:
1. Update counts: [1, 1]
2. Exception: `AlreadyExists`

In other words: The first two statements succeeded, the third failed. The current retry logic does not appear to take this into account.

Note that the following could in theory also happen:

```sql
insert into test (id, value) values (1, 'One');
insert into test (id, value) values (2, 'Two');
insert into test (id, value) values (3, 'Three');
```

Result:
1. Update counts: [1, 1]
2. Exception: `Aborted`

This means that the transaction was aborted after executing the first two DML statements.

Reply:
As discussed created a bug https://buganizer.corp.google.com/issues/319404848 for this and will take it later as part of separate PR

Reply:
ACK